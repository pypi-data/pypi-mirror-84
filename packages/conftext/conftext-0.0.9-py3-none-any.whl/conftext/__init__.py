__version__ = "0.0.9"

from configparser import ConfigParser
from pydantic import BaseModel
from invoke import task, Program, Collection
from conftext import conftext
from conftext import conf_ini


###
# Schema
###

# NOTE: ConfigParser doesn't like `None`, so we use strings for now.

class MultiTenantV1(BaseModel):
    service: str = str(None)
    context: str = str(None)


class MultiTenantV2(BaseModel):
    tenant: str = str(None)
    environment: str = str(None)


###
# Public API
###

class NoConftext(Exception):
    pass


def get_section(conftext, section_name=None):
    """
    Select default or named section in conftext config. Named sections will inherit values from
    defaults for options that have not been set.
    """
    if section_name and section_name in conftext:
        return conftext[section_name]
    else:
        return conftext.defaults()


def get_config(**kwargs) -> ConfigParser:
    """
    Get config
    
    kwargs can be used to overwrite settings from config file. If the value of a kwarg is `None`,
    it will be ignored.
    """
    config_file = conftext.find_file()
    
    if not config_file and not kwargs:
        raise NoConftext('No "%s" file found and no kwargs given.' % conftext.CONFTEXT_FILENAME)
    
    if config_file:
        config = get_section(conftext.read_from_file(config_file), conftext.CONFTEXT_SECTION)
    else:
        config = dict()
    
    for key, val in kwargs.items():
        if val is not None:
            config[key] = val
    
    return config


def get_config_v2(**kwargs) -> ConfigParser:
    config_file = conftext.find_file()
    
    if not config_file and not kwargs:
        raise NoConftext('No "%s" file found and no kwargs given.' % conftext.CONFTEXT_FILENAME)
    
    config = conftext.read_from_file(config_file)
    
    for key, val in kwargs.items():
        if val is not None:
            config[config.default_section][key] = val
    
    return config


def get_ini_config(config_filepath, conftext_section=None):
    """
    Get config from ini file
    
    Can use module name to look for config file in corresponding path under `~/.config`. Only use
    the conftext machinery if its nexessary to select from multiple sections in config file.
    """
    config_file = conf_ini.read_config(config_filepath)
    return conf_ini.get_config_section(config_file, conftext_section)


###
# Tasks
###

@task(default=True)
def show(ctxt, verbose=False):
    """
    Show config context
    """
    config_file = conftext.find_file()
    if not config_file:
        config_file = conftext.ask_path()
        config = conftext.initialize_config()
        conftext.write_to_file(config_file, config)
    if verbose:
        print(config_file)
    print(conftext.format_string(conftext.read_from_file(config_file)))


@task
def schemas(ctxt):
    """
    List all schemas
    """
    for key, val in conftext.get_schemas().items():
        print(key, val.schema_json())


@task
def init(ctxt):
    """
    Initialize a conftext file
    """
    config = conftext.initialize_config()
    print(config)


###
# Invoke setup
###

namespace = Collection()
namespace.add_task(init)
namespace.add_task(show)
namespace.add_task(schemas)
program = Program(namespace=namespace)
