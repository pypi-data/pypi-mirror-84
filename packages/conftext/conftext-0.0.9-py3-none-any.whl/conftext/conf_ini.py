import os
import configparser


def read_config(filepath):
    if os.path.isfile(filepath):
        config = configparser.ConfigParser()
        config.read(filepath)
        return config
    raise FileNotFoundError(filepath)


def get_config_section(config_file, conftext_section):
    """
    Get config section
    
    First check the loaded config to see if we have more than one section to choose from. Only if we
    do, we move on to use conftext to select the appropriate config section.
    """
    section_name = None
    
    if config_file.sections():
        
        if conftext_section:
            # We select section based on two hard-coded dimmensions for now; service and context. This
            # should be based on the config schema.
            if "service" in conftext_section and conftext_section["service"] in config_file:
                section_name = conftext_section["service"]
            elif "context" in conftext_section and conftext_section["context"] in config_file:
                section_name = conftext_section["context"]
            else:
                print(f"WARNING: No match for {conftext_section} in {config_file}")
        else:
            print(
                f"WARNING: No conftext_section ({conftext_section}) for selecting among sections "
                f"({config_file._sections}) in config file {config_file}."
            )
    
    elif config_file.defaults():
        section_name = configparser.DEFAULTSECT
    
    if not section_name:
        raise configparser.NoSectionError(section_name)
    
    return config_file[section_name]
