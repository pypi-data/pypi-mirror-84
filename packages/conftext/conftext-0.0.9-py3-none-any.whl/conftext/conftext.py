import os
import sys
import pkg_resources
from configparser import ConfigParser

CONFTEXT_FILENAME = '.conftext.ini'
CONFTEXT_SECTION = 'conftext'


def user_home():
    return os.path.expanduser('~')


def default_config_path():
    return os.path.join(user_home(), '.config', CONFTEXT_FILENAME[1:])


stop_paths = [os.path.sep, os.path.dirname(user_home())]


def find_file():
    """
    Find config files
    
    Will first look for `.conftext.ini` in the current working dir and then subsequently upwards the
    users home directory, or if home is not found, upwards to top-level dir `/`.
    
    Then it will look for `~/.config/conftext.ini`
    
    Returns `None` if no files found.
    """
    current_path = os.getcwd()
    
    # Check from current dir upwards `stop_paths` dir for `.conftext.ini`
    while current_path not in stop_paths:
        candidate_path = os.path.join(current_path, CONFTEXT_FILENAME)
        if os.path.isfile(candidate_path):
            return candidate_path
        else:
            current_path = os.path.dirname(current_path)
    
    # Check for default path `~/.config/conftext.ini`
    candidate_path = default_config_path()
    if os.path.isfile(candidate_path):
        return candidate_path
    
    # Finally return None if no conf file found.
    return None


def ask_path():
    cwd_config_path = os.path.join(os.getcwd(), CONFTEXT_FILENAME)
    print("No config file was found. Create one with default settings?")
    print("[1] %s" % default_config_path())
    print("[2] %s" % cwd_config_path)
    choice = input("Type 1 or 2 then enter: ")
    
    if choice == '1':
        return default_config_path()
    elif choice == '2':
        return cwd_config_path
    else:
        sys.exit('Invalid choice: %s' % choice)


def write_to_file(filepath, config):
    with open(filepath, 'w') as config_file_obj:
        config.write(config_file_obj)
    print("Wrote config to %s" % filepath)


def read_from_file(config_filepath_or_string: str) -> ConfigParser:
    config = ConfigParser(default_section=CONFTEXT_SECTION)
    if config_filepath_or_string:
        if not config.read(config_filepath_or_string):
            config.read_string(config_filepath_or_string)
    return config


def format_string(config):
    return '[%s]' % '/'.join(config[CONFTEXT_SECTION].values())


def get_schemas():
    return {ep.name: ep.load() for ep in pkg_resources.iter_entry_points(group='conftext')}


def initialize_config() -> ConfigParser:
    """
    Create initial config
    """
    
    # select from registered schamas in package entries
    schemas = dict(enumerate(get_schemas().items(), start=1))
    print("Schemas registered as entry points:")
    for num, (name, schema) in schemas.items():
        print(num, name, schema)
    
    in_val = input('select schema by number: ')
    name, selected_schema = schemas[int(in_val)]
    print("selected '%s': %s" % (name, selected_schema))
    
    print(dict(selected_schema()))
    
    # create and return selected config
    schema = ConfigParser()
    schema[CONFTEXT_SECTION] = dict(selected_schema())
    return schema
