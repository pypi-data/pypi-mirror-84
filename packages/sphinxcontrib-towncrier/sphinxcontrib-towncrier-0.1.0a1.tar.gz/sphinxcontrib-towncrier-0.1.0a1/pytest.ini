[pytest]
addopts =
    # `pytest-xdist`:
    -n auto

    --durations=10
    -v
    -ra
    # Show local variables in tracebacks
    --showlocals
    # https://docs.pytest.org/en/stable/doctest.html
    --doctest-modules
    --junitxml=.test-results/pytest/results.xml

    # Fail on non-existing markers:
    --strict
    --strict-markers

doctest_optionflags = ALLOW_UNICODE ELLIPSIS
# Marks tests with an empty parameterset as xfail(run=False)
empty_parameter_set_mark = xfail
faulthandler_timeout = 30

# https://docs.pytest.org/en/stable/usage.html#creating-junitxml-format-files
junit_duration_report = call
junit_family = xunit2
junit_logging = all
junit_log_passing_tests = true
junit_suite_name = towncrier_test_suite

minversion = 6.1.0

norecursedirs =
    build
    dist
    docs
    src/sphinxcontrib_towncrier.egg-info
    .cache
    .eggs
    .git
    .github
    .tox
    *.egg
testpaths = tests/
xfail_strict = true
