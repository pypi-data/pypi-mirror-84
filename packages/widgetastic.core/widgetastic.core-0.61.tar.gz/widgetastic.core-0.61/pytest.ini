[pytest]
addopts =
    -v
    --cov widgetastic
    --cov-report term-missing
    --cov-report xml
    --no-cov-on-fail
