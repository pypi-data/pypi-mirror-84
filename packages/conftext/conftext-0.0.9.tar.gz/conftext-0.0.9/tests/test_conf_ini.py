import pytest
from conftext import conftext
from conftext import get_ini_config
from conftext import conf_ini

CONFTEXT_ONLY_GLOBAL = conftext.read_from_file("""
[conftext]
context = development
service = dummy
""")

CONFTEXT_WITH_MODULE_SECTION = conftext.read_from_file("""
[conftext]
context = development
service = dummy

[package.module]
context = development
""")

FILEPATH_DEFAULT_SECTION = "tests/config/package/module/default_section.ini"
FILEPATH_MULTI_SECTION = "tests/config/package/module/multi_section.ini"


def test_get_ini_config_ok(regtest):
    print(get_ini_config(FILEPATH_DEFAULT_SECTION), file=regtest)


@pytest.mark.xfail(raises=FileNotFoundError)
def test_get_ini_config_nok():
    get_ini_config("./invalid/path.ini")


def test_get_config_section(regtest):
    
    config = conf_ini.read_config(FILEPATH_DEFAULT_SECTION)
    print(conf_ini.get_config_section(
        config, CONFTEXT_WITH_MODULE_SECTION["package.module"]), file=regtest)
    print(conf_ini.get_config_section(
        config, CONFTEXT_ONLY_GLOBAL.defaults()), file=regtest)
    
    config = conf_ini.read_config(FILEPATH_MULTI_SECTION)
    print(conf_ini.get_config_section(
        config, CONFTEXT_WITH_MODULE_SECTION["package.module"]), file=regtest)
    print(conf_ini.get_config_section(
        config, CONFTEXT_ONLY_GLOBAL.defaults()), file=regtest)


def test_get_config(regtest):
    print(get_ini_config(
        FILEPATH_MULTI_SECTION,
        CONFTEXT_WITH_MODULE_SECTION["package.module"]
    ), file=regtest)
