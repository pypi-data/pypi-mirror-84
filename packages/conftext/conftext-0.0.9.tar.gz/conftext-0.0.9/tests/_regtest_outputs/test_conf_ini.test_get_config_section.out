<Section: DEFAULT>
<Section: DEFAULT>
<Section: development>
<Section: development>
